src/mame/misc/fruitctk.cpp

```diff
@@ -0,0 +1,316 @@
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_UNKNOWN )
+    PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_UNKNOWN )
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_CUSTOM) PORT_READ_LINE_DEVICE_MEMBER("hopper", FUNC(hopper_device::line_r))  // is from hopper? will cause error 02 During reset nvram
+    PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_UNKNOWN ) 
+    PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_UNKNOWN ) 
+
+
+
+PORT_START("DSW1")
+	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW1:1")
+	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW1:2")
+	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW1:3")
+	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW1:4")
+	PORT_DIPUNKNOWN_DIPLOC(0x10, 0x10, "SW1:5")
+	PORT_DIPUNKNOWN_DIPLOC(0x20, 0x20, "SW1:6")
+	PORT_DIPUNKNOWN_DIPLOC(0x40, 0x40, "SW1:7")
+	PORT_DIPUNKNOWN_DIPLOC(0x80, 0x80, "SW1:8")
+
+
+PORT_START("DSW2")
+	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW2:1")
+	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW2:2")
+	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW2:3")
+	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW2:4")
+INPUT_PORTS_END
+
+
+
+
+void fruitctk_state::p1_w(uint8_t data)
+{
+
+	m_hopper->motor_w(BIT(data, 3)); // ?? from guessed. Wrong
+    logerror("P1.0 Write to %02x\n", data);
+
+	
+}
+
+
+
+void fruitctk_state::multiplex_7seg_w(uint8_t data)
+{
+	m_selected_7seg_module = data;
+}
+
+uint8_t fruitctk_state::keyboard_r()
+{
+	switch (m_selected_7seg_module & 0x07)
+	{
+	case 0:
+	case 1:
+	case 2:
+	case 3:
+		return m_inputs[m_selected_7seg_module & 0x07]->read();
+	default:
+		return 0x00;
+	}
+}
+
+void fruitctk_state::display_7seg_data_w(uint8_t data)
+{
+	static const uint8_t patterns[16] = { 0x3f, 0x06, 0x5b, 0x4f, 0x66, 0x6d, 0x7c, 0x07, 0x7f, 0x67, 0, 0, 0, 0, 0, 0 }; // Code was from marywu to decode 7 segment display.
+
+	m_digits[2 * m_selected_7seg_module + 0] = patterns[data & 0x0f];
+	m_digits[2 * m_selected_7seg_module + 1] = patterns[data >> 4];
+}
+
+void fruitctk_state::program_map(address_map &map)
+{
+	map(0x0000, 0xffff).rom();
+}
+
+void fruitctk_state::io_map(address_map &map)
+{
+map(0x8000, 0x87ff).ram().share("nvram");
+map(0xc000, 0xc001).w("opll", FUNC(ym2413_device::write));
+map(0x9002, 0x9003).rw("ay1", FUNC(ay8910_device::data_r), FUNC(ay8910_device::address_data_w));
+map(0xb000, 0xb001).rw("i8279", FUNC(i8279_device::read), FUNC(i8279_device::write));
+map(0xb004, 0xb004).rw(m_oki, FUNC(okim6295_device::read), FUNC(okim6295_device::write));
+map(0xb008, 0xb00a).rw("ppi1", FUNC(i8255_device::read), FUNC(i8255_device::write)); // ???
+// map(0xb00b, 0xb00b).nopw(); // ?? Write = 80
+// map(0xb00c, 0xb00c).nopw(); // Led
+map(0xb00d, 0xb00f).rw("ppi2", FUNC(i8255_device::read), FUNC(i8255_device::write)); // ???
+}
+
+void fruitctk_state::machine_start()
+{
+	m_digits.resolve();
+	m_leds.resolve();	
+}
+
+void fruitctk_state::fruitctk(machine_config &config)
+{
+	/* Video */
+    config.set_default_layout(layout_marywu);
+	
+	/* basic machine hardware */
+	i8051_device &maincpu(I8051(config, "maincpu", XTAL(12'000'000)));

```

Comment:
Conventionally the machine_config should have following order:
```c++
void config_state::config_name(machine_config &config)
{
    [CPU declarations]
    [I/O class devices]
    [video]
    [sound]
}
```

Put in another way, `config.set_default_layout` should go before sound, in no case as first declared thing.